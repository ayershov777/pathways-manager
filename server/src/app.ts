import express from "express";
import cors from "cors";
import morgan from "morgan";
import { router as pathwaysRouter } from "./routes/pathways.routes";
import { routeNotFoundMiddleware, errorHandlerMiddleware } from "./middleware/error.middleware";

/**
 * Create and configure Express application
 */
const createApp = () => {
    // Initialize Express app
    const app = express();

    // Apply middleware
    app.use(morgan('dev'));
    app.use(cors());
    app.use(express.json());

    // Mount routes
    app.use('/api/v1/pathways', pathwaysRouter);

    // 404 handler - must come after all routes
    app.use(routeNotFoundMiddleware);

    // Global error handler - must be the last middleware
    app.use(errorHandlerMiddleware);

    return app;
};

// Create the application
const app = createApp();

export default app;