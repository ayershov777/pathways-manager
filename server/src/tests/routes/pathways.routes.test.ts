import request from 'supertest';
import mongoose from 'mongoose';
import app from '../../app';
import { Pathway } from '../../models/pathway.model';
import { User } from '../../models/user.model';
import jwt from 'jsonwebtoken';

describe('Pathway Routes', () => {
    let pathwayId: string;
    let userId: string;
    let anotherUserId: string;
    let userToken: string;
    let anotherUserToken: string;

    beforeEach(async () => {
        // Create test users
        const user = await User.create({
            email: 'user@example.com',
            password: 'password123',
            name: 'Test User'
        });
        userId = user._id.toString();
        userToken = jwt.sign({ id: userId }, process.env.JWT_SECRET || 'test-secret');

        const anotherUser = await User.create({
            email: 'another@example.com',
            password: 'password123',
            name: 'Another User'
        });
        anotherUserId = anotherUser._id.toString();
        anotherUserToken = jwt.sign({ id: anotherUserId }, process.env.JWT_SECRET || 'test-secret');

        // Create a test pathway owned by user
        const pathway = await Pathway.create({
            name: 'Test Pathway',
            description: 'Test Description',
            owner: user._id
        });
        pathwayId = pathway._id.toString();

        // Create additional pathways for pagination tests
        const pathways = [];
        for (let i = 1; i <= 15; i++) {
            pathways.push({
                name: `Pathway ${i}`,
                description: `Description ${i}`,
                owner: i % 2 === 0 ? user._id : anotherUser._id
            });
        }
        await Pathway.insertMany(pathways);
    });

    describe('GET /api/v1/pathways', () => {
        it('should get all pathways with default pagination without authentication', async () => {
            const response = await request(app)
                .get('/api/v1/pathways')
                .expect(200);

            expect(response.body).toMatchSnapshot();
        });

        it('should get all pathways with default pagination with authentication', async () => {
            const response = await request(app)
                .get('/api/v1/pathways')
                .set('Authorization', `Bearer ${userToken}`)
                .expect(200);

            expect(response.body).toMatchSnapshot();
        });

        it('should get pathways with custom pagination', async () => {
            const response = await request(app)
                .get('/api/v1/pathways?page=2&limit=5')
                .expect(200);

            expect(response.body).toMatchSnapshot();
        });
    });

    describe('GET /api/v1/pathways/:id', () => {
        it('should get a pathway by id without authentication', async () => {
            const response = await request(app)
                .get(`/api/v1/pathways/${pathwayId}`)
                .expect(200);

            expect(response.body).toMatchSnapshot();
        });

        it('should get a pathway by id with authentication', async () => {
            const response = await request(app)
                .get(`/api/v1/pathways/${pathwayId}`)
                .set('Authorization', `Bearer ${userToken}`)
                .expect(200);

            expect(response.body).toMatchSnapshot();
        });

        it('should return 404 for non-existent pathway', async () => {
            const nonExistentId = new mongoose.Types.ObjectId().toString();

            const response = await request(app)
                .get(`/api/v1/pathways/${nonExistentId}`)
                .expect(404);

            expect(response.body).toMatchSnapshot();
        });

        it('should return 400 for invalid id format', async () => {
            const response = await request(app)
                .get('/api/v1/pathways/invalid-id')
                .expect(400);

            expect(response.body).toMatchSnapshot();
        });
    });

    describe('POST /api/v1/pathways', () => {
        it('should create a new pathway when authenticated', async () => {
            const newPathway = {
                name: 'New Pathway',
                description: 'New Description'
            };

            const response = await request(app)
                .post('/api/v1/pathways')
                .set('Authorization', `Bearer ${userToken}`)
                .send({ pathway: newPathway })
                .expect(201);

            expect(response.body).toMatchSnapshot();

            // Verify it was created in the database
            const savedPathway = await Pathway.findById(response.body.pathway._id);
            expect(savedPathway).toBeDefined();
            expect(savedPathway?.name).toBe(newPathway.name);
        });

        it('should return 401 when not authenticated', async () => {
            const newPathway = {
                name: 'New Pathway',
                description: 'New Description'
            };

            const response = await request(app)
                .post('/api/v1/pathways')
                .send({ pathway: newPathway })
                .expect(401);

            expect(response.body).toMatchSnapshot();
        });

        it('should return 400 if required fields are missing', async () => {
            const response = await request(app)
                .post('/api/v1/pathways')
                .set('Authorization', `Bearer ${userToken}`)
                .send({ pathway: { name: 'Missing Description' } })
                .expect(400);

            expect(response.body).toMatchSnapshot();
        });
    });

    describe('PATCH /api/v1/pathways/:id', () => {
        it('should update pathway when owner is authenticated', async () => {
            const updateData = {
                name: 'Updated Pathway',
                description: 'Updated Description'
            };

            const response = await request(app)
                .patch(`/api/v1/pathways/${pathwayId}`)
                .set('Authorization', `Bearer ${userToken}`)
                .send({ pathway: updateData })
                .expect(200);

            expect(response.body).toMatchSnapshot();

            // Verify it was updated in the database
            const updatedPathway = await Pathway.findById(pathwayId);
            expect(updatedPathway?.name).toBe(updateData.name);
        });

        it('should return 401 when not authenticated', async () => {
            const updateData = {
                name: 'Updated Pathway',
                description: 'Updated Description'
            };

            const response = await request(app)
                .patch(`/api/v1/pathways/${pathwayId}`)
                .send({ pathway: updateData })
                .expect(401);

            expect(response.body).toMatchSnapshot();
        });

        it('should return 403 when not the owner', async () => {
            const updateData = {
                name: 'Updated by Non-Owner',
                description: 'Should Fail'
            };

            const response = await request(app)
                .patch(`/api/v1/pathways/${pathwayId}`)
                .set('Authorization', `Bearer ${anotherUserToken}`)
                .send({ pathway: updateData })
                .expect(403);

            expect(response.body).toMatchSnapshot();
        });

        it('should return 404 for non-existent pathway', async () => {
            const nonExistentId = new mongoose.Types.ObjectId().toString();

            const response = await request(app)
                .patch(`/api/v1/pathways/${nonExistentId}`)
                .set('Authorization', `Bearer ${userToken}`)
                .send({ pathway: { name: 'Updated Name' } })
                .expect(404);

            expect(response.body).toMatchSnapshot();
        });
    });

    describe('DELETE /api/v1/pathways/:id', () => {
        it('should delete pathway when owner is authenticated', async () => {
            const response = await request(app)
                .delete(`/api/v1/pathways/${pathwayId}`)
                .set('Authorization', `Bearer ${userToken}`)
                .expect(200);

            expect(response.body).toMatchSnapshot();

            // Verify it was deleted from the database
            const deletedPathway = await Pathway.findById(pathwayId);
            expect(deletedPathway).toBeNull();
        });

        it('should return 401 when not authenticated', async () => {
            const response = await request(app)
                .delete(`/api/v1/pathways/${pathwayId}`)
                .expect(401);

            expect(response.body).toMatchSnapshot();
        });

        it('should return 403 when not the owner', async () => {
            const response = await request(app)
                .delete(`/api/v1/pathways/${pathwayId}`)
                .set('Authorization', `Bearer ${anotherUserToken}`)
                .expect(403);

            expect(response.body).toMatchSnapshot();
        });

        it('should return 404 for non-existent pathway', async () => {
            const nonExistentId = new mongoose.Types.ObjectId().toString();

            const response = await request(app)
                .delete(`/api/v1/pathways/${nonExistentId}`)
                .set('Authorization', `Bearer ${userToken}`)
                .expect(404);

            expect(response.body).toMatchSnapshot();
        });
    });
});