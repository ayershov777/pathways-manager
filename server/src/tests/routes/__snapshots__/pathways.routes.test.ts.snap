// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pathway Routes DELETE /api/v1/pathways/:id should delete an existing pathway 1`] = `
{
  "message": "Pathway successfully deleted",
  "pathway": {
    "__v": 0,
    "_id": "6808546d3ecabe44f5215e4f",
    "description": "Test Description",
    "modules": [],
    "name": "Test Pathway",
  },
}
`;

exports[`Pathway Routes DELETE /api/v1/pathways/:id should return 404 for non-existent pathway 1`] = `
{
  "error": {
    "message": "Pathway not found",
  },
  "success": false,
}
`;

exports[`Pathway Routes GET /api/v1/pathways should get all pathways with default pagination 1`] = `
{
  "pagination": {
    "page": 1,
    "pages": 2,
    "total": 16,
  },
  "pathways": [
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d57",
      "description": "Test Description",
      "name": "Test Pathway",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d5a",
      "description": "Description 1",
      "name": "Pathway 1",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d5b",
      "description": "Description 2",
      "name": "Pathway 2",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d5c",
      "description": "Description 3",
      "name": "Pathway 3",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d5d",
      "description": "Description 4",
      "name": "Pathway 4",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d5e",
      "description": "Description 5",
      "name": "Pathway 5",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d5f",
      "description": "Description 6",
      "name": "Pathway 6",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d60",
      "description": "Description 7",
      "name": "Pathway 7",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d61",
      "description": "Description 8",
      "name": "Pathway 8",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d62",
      "description": "Description 9",
      "name": "Pathway 9",
    },
  ],
}
`;

exports[`Pathway Routes GET /api/v1/pathways should get pathways with custom pagination 1`] = `
{
  "pagination": {
    "page": 2,
    "pages": 4,
    "total": 16,
  },
  "pathways": [
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d73",
      "description": "Description 5",
      "name": "Pathway 5",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d74",
      "description": "Description 6",
      "name": "Pathway 6",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d75",
      "description": "Description 7",
      "name": "Pathway 7",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d76",
      "description": "Description 8",
      "name": "Pathway 8",
    },
    {
      "__v": 0,
      "_id": "6808546d3ecabe44f5215d77",
      "description": "Description 9",
      "name": "Pathway 9",
    },
  ],
}
`;

exports[`Pathway Routes GET /api/v1/pathways/:id should get a pathway by id 1`] = `
{
  "pathway": {
    "__v": 0,
    "_id": "6808546d3ecabe44f5215d82",
    "description": "Test Description",
    "modules": [],
    "name": "Test Pathway",
  },
}
`;

exports[`Pathway Routes GET /api/v1/pathways/:id should return 400 for invalid id format 1`] = `
{
  "error": {
    "message": "Invalid pathway ID format",
  },
  "success": false,
}
`;

exports[`Pathway Routes GET /api/v1/pathways/:id should return 404 for non-existent pathway 1`] = `
{
  "error": {
    "message": "Pathway not found",
  },
  "success": false,
}
`;

exports[`Pathway Routes PATCH /api/v1/pathways/:id should return 400 if pathway object is missing 1`] = `
{
  "error": {
    "message": "Request body must contain a pathway object",
  },
  "success": false,
}
`;

exports[`Pathway Routes PATCH /api/v1/pathways/:id should return 404 for non-existent pathway 1`] = `
{
  "error": {
    "message": "Pathway not found",
  },
  "success": false,
}
`;

exports[`Pathway Routes PATCH /api/v1/pathways/:id should update an existing pathway 1`] = `
{
  "pathway": {
    "__v": 0,
    "_id": "6808546d3ecabe44f5215e0f",
    "description": "Updated Description",
    "modules": [],
    "name": "Updated Pathway",
  },
}
`;

exports[`Pathway Routes POST /api/v1/pathways should create a new pathway 1`] = `
{
  "pathway": {
    "__v": 0,
    "_id": "6808546d3ecabe44f5215dd1",
    "description": "New Description",
    "modules": [],
    "name": "New Pathway",
  },
}
`;

exports[`Pathway Routes POST /api/v1/pathways should return 400 if pathway object is missing 1`] = `
{
  "error": {
    "message": "Request body must contain a pathway object",
  },
  "success": false,
}
`;

exports[`Pathway Routes POST /api/v1/pathways should return 400 if required fields are missing 1`] = `
{
  "error": {
    "message": "Validation error: description: must have required property 'description'",
  },
  "success": false,
}
`;

exports[`Pathway Routes POST /api/v1/pathways should return 409 if pathway name already exists 1`] = `
{
  "error": {
    "message": "The specified pathway name is unavailable.",
  },
  "success": false,
}
`;
